x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -493,131 +549,139 @@ public void runAsDenied(Authentication authentication, String action, TransportM
+                    .with(PRINCIPAL_RUN_BY_FIELD_NAME, authentication.getUser().authenticatedUser().principal())
+                    .with(PRINCIPAL_RUN_BY_REALM_FIELD_NAME, authentication.getAuthenticatedBy().getName());
+        } else {
+            logEntry.with(PRINCIPAL_REALM_FIELD_NAME, authentication.getAuthenticatedBy().getName());
         }
-        sb.append(authentication.getAuthenticatedBy().getName()).append("]");
-        return sb.toString();
-    }
-
-    private static String hostAttributes(RestRequest request) {
-        final InetSocketAddress socketAddress = request.getHttpChannel().getRemoteAddress();
-        String formattedAddress = NetworkAddress.format(socketAddress.getAddress());
-        return "origin_address=[" + formattedAddress + "]";
     }
 
-    protected static String originAttributes(ThreadContext threadContext, TransportMessage message, LocalNodeInfo localNodeInfo) {
-        return restOriginTag(threadContext).orElse(transportOriginTag(message).orElse(localNodeInfo.localOriginTag));
+    private static void opaqueId(ThreadContext threadContext, StringMapMessage logEntry) {
+        final String opaqueId = threadContext.getHeader(Task.X_OPAQUE_ID);
+        if (opaqueId != null) {
+            logEntry.with(OPAQUE_ID_FIELD_NAME, opaqueId);
+        }
     }
 
-    private String opaqueId() {
-        String opaqueId = threadContext.getHeader(Task.X_OPAQUE_ID);
-        if (opaqueId != null) {
-            return ", opaque_id=[" + opaqueId + "]";
-        } else {
-            return "";
+    private static void restOrigin(RestRequest request, StringMapMessage logEntry) {
+        assert LOCAL_ORIGIN_FIELD_VALUE.equals(logEntry.get(ORIGIN_TYPE_FIELD_NAME)); // this is the default
+        final InetSocketAddress socketAddress = request.getHttpChannel().getRemoteAddress();
+        if (socketAddress != null) {
+            final String formattedAddress = NetworkAddress.format(socketAddress.getAddress());
+            logEntry.with(ORIGIN_TYPE_FIELD_NAME, REST_ORIGIN_FIELD_VALUE).with(ORIGIN_ADDRESS_FIELD_NAME, formattedAddress);
         }
+        // fall through to local_node default
     }
 
-    private static Optional<String> restOriginTag(ThreadContext threadContext) {
-        final InetSocketAddress restAddress = RemoteHostHeader.restRemoteAddress(threadContext);
-        if (restAddress == null) {
-            return Optional.empty();
+    private static void restUri(RestRequest request, StringMapMessage logEntry) {
+        if (Strings.hasLength(request.path())) {
+            logEntry.with(URL_PATH_FIELD_NAME, request.path());
+        }
+        if (request.params().isEmpty() == false) {
+            // TODO this is probably wrong, params are decoded and they might contain the delimiter
+            logEntry.with(URL_QUERY_FIELD_NAME, Strings.collectionToDelimitedString(request.params().entrySet(), "&"));
```

Comment:
Previously there was a field named `uri` that included the query string. To make it more compliant to [ecs-url-fields]( https://github.com/elastic/ecs#-url-fields) I have split it to `url.path` and `url.query`. What do you think @ruflin ?
However, `url.query` is potentially ambiguous because URL parameters are decoded at this point so they might contain the `&` character.

Reply:
In ECS `url.query` is also just a string and not an array but it's analyzed so Elasticsearch would do the tokenisation for us.

Reply:
okay. I presume it's also decoded, `%26` -> `&`? The trouble is that I am doing concatenation with a character that might also be present in the value (`&`) when decoded.
I am fine with it as it is, I was just raising a potential contestable topic.

Reply:
We don't have any definitions yet on ECS for the encoding. If `&` is a the value, what will your query string look like?

Reply:
> If & is a the value, what will your query string look like?

What I meant here is that the query string parts are decoded such that there are no more % encoded values. Consequently, if a query param or value had `%26` in it, it gets decoded to `&` which is also the separator of individual `param=value` pairs.

Anyway, this is not relevant anymore. I have looked at the ECS schema, and since there is provisioning about encoding - as you said - and since it is expecting the whole query string as text, I suppose the analyzer expects encoded values since otherwise decoded string will be ambiguous - `=` and `&` are special characters they **have** to be encoded.

The outcome is that `url.query` is what follows `?` and until `#`, as is, in the URL. I fathom this is ECS "compliant", isn't it?